### PR TITLE
Replace patches with commits

### PR DESCRIPTION
**Pull request by erick-xanadu**
**Context:** Instead of using patches, commit the generated files into the repository. Please note that this just shifts the maintance a bit. Instead of applying patches and not having to worry about when the ODS definition changes now we are not applying patches and have to manually decide when to copy the generated files into the repository (when there's a mismatch in the ODS definition that generated the files and the ODS definition when building the generated files). There will be no patches applied automatically, but the changes will be committed to the repository. 

I still kept the patches because when the ODS changes, then we need to copy paste something into the repository. Copy pasting would change back the python bindings to their original state. We would have to manually write them back and look for changes. The application of patches are optional but I suspect that people who have to make ODS changes will find them beneficial like I just did.

EDIT: This ticket was also used as a discussion place for generating generic operations from the MLIR bindings.

** Ticket: ** [sc-41540]

### CODE REVIEW

**Comment by pengmai** [comment]
As a word of caution, [relying on generic builder APIs is explicitly discouraged in the MLIR docs](https://mlir.llvm.org/docs/Bindings/Python/#:~:text=This%20form%20is%20discouraged%20from%20use). The risk here is that ODS operand ordering is not a guarantee, and at any point someone can change the operand ordering upstream. This can cause silent failures or incorrect results in unexpected ways. [See also this discussion on a somewhat similar case](https://github.com/EnzymeAD/Enzyme/pull/1134#discussion_r1219437474).

Are there instances where the inferred result type of a `tensor.extract` op are incorrect?

**Comment by erick-xanadu** [comment]
@pengmai I agree 100% with you. I highlighted the MLIR documentation in the notion document regarding getting rid of these patches. If there's another way to get rid of patches and use the python bindings I am all up for it, but I am happy this discussion is taking place.

As a note, we need the patches because JAX does not ship with the bindings for some dialects and therefore the Python/C++ extension module also lacks some of the functionality (like `InferTypeOpInterface`).

**Comment by pengmai** [comment]
> As a note, we need the patches because JAX does not ship with the bindings for some dialects and therefore the Python/C++ extension module also lacks some of the functionality (like `InferTypeOpInterface` for those dialects).

I see, that's super unfortunate 😢 As long as we're aware of the risks, I trust the team's decision. The patches are definitely an unwieldy part of the build process.

**Comment by erick-xanadu** [comment]
Just as a note here, I have changed my mind and I think the approach with patches is better than this one. Happy to leave this PR as a draft so that other people can take a stab at this problem. Feel free to discuss the advantages and disadvantages of each.

**Comment by erick-xanadu** [comment]
Anyone knows how to disable code factor for a directory?

**Comment by dime10** [comment]
I'm a bit concerned that we are checking-in too many files unnecessarily. I think we should try to include only what is absolutely necessary to reduce the maintenance burden.
For example:
- we mentioned arith could be replaced with mhlo ops, eliminating the need for its binding file
- why is ods_common needed? at least in the past we didn't include it
- do we use sparse tensors now?
- the runtime module doesn't need to be checked-in as we are using it as is right?

Having said, maybe there is a good reason to include as much as you did?

**Comment by dime10** [comment]
Another approach might be to reduce the ODS files even further, with only the op definitions that we actually use. I'm not sure if that is more or less work however 🤔

**Comment by erick-xanadu** [comment]
> I'm a bit concerned that we are checking-in too many files unnecessarily. I think we should try to include only what is absolutely necessary to reduce the maintenance burden.

I used the same procedure when copying files during the wheel creation, so these files are exactly the same that are shipped in the wheel.

> we mentioned arith could be replaced with mhlo ops, eliminating the need for its binding file

I'll correct this a bit, arith operations could be replaced with StableHLO ops. Yes, we could do that and I agree that that is a good thing. The only comment I have is that we will still be required to keep the tensor bindings since StableHLO operates on tensors and not scalars, so if our operations expect scalar arguments we will need to get them. Also potentially cast them. See [here for an idea of what removing the arithmetic operations would look like.](https://github.com/PennyLaneAI/catalyst/pull/185/commits/cc1f40c9d257571b585cf5da5a29b4930a738e4b)

> why is ods_common needed? at least in the past we didn't include it

You can see in the build wheel process that we are indeed copying `_ods_common`, I believe it is necessary but I will double check. You are correct that we don't need sparse tensors. These get copy accidentally because they just match `_tensor`. 

> the runtime module doesn't need to be checked-in as we are using it as is right?

If we don't check it in, we will need to copy paste it during wheel creation. It just feels like having everything in the repository as opposed to "outside the build system" is the overall strategy here. As one can argue that the patches are an inconvenience when building from source, one can argue that copying these files are an inconvenience when building the wheel from source.

> Another approach might be to reduce the ODS files even further, with only the op definitions (convenience functions around generic op creation) that we actually use. I'm not sure if that is more or less work however

I prefer to keep things as close to the generated files as possible, however I agree that reducing our dependence on them (like substituting arithmetic operations with stable hlo) is a good approach. In my opinion, if we use one single op we should just include the whole dialect files.

**Comment by dime10** [comment]
> I used the same procedure when copying files during the wheel creation, so these files are exactly the same that are shipped in the wheel.

Ah I see, that explains why all these additional files are being checked in.

IMO, build artifacts should rarely if ever be checked into the repository (placing them into the wheel is fine of course since that is the purpose of distributing a pre-built project). The one exception, as is the case here, is that we have to manually modify some of them. So to me any file that doesn't need to be modified or does not need to be included as a result of the modifications should be checked into the repository.

> If we don't check it in, we will need to copy paste it during wheel creation. It just feels like having everything in the repository as opposed to "outside the build system" is the overall strategy here. As one can argue that the patches are an inconvenience when building from source, one can argue that copying these files are an inconvenience when building the wheel from source.

I think I disagree here because copying build artifacts into a single location is the whole purpose of a packaging process.

> I prefer to keep things as close to the generated files as possible, however I agree that reducing our dependence on them (like substituting arithmetic operations with stable hlo) is a good approach. In my opinion, if we use one single op we should just include the whole dialect files.

Sounds good, we can keep the files as a whole but substitute as much of the dialects as possible with stable HLO. Currently that just includes `arith` I believe, since as you mention we do need `tensor.extract` and we also explicitly need the `scf` ops, but for example our use of `tensor.from_elements` might be able to be replaced. Anything to reduce the "surface" of these files would be good in my eyes.

**Comment by erick-xanadu** [comment]
> > If we don't check it in, we will need to copy paste it during wheel creation. It just feels like having everything in the repository as opposed to "outside the build system" is the overall strategy here. As one can argue that the patches are an inconvenience when building from source, one can argue that copying these files are an inconvenience when building the wheel from source.

> I think I disagree here because copying build artifacts into a single location is the whole purpose of a packaging process.

@maliasadi what do you think about this point, specially considering building wheels directly from the python?

**Comment by maliasadi** [comment]
> > > If we don't check it in, we will need to copy paste it during wheel creation. It just feels like having everything in the repository as opposed to "outside the build system" is the overall strategy here. As one can argue that the patches are an inconvenience when building from source, one can argue that copying these files are an inconvenience when building the wheel from source.
> 
> > I think I disagree here because copying build artifacts into a single location is the whole purpose of a packaging process.
> 
> @maliasadi what do you think about this point, specially considering building wheels directly from the python?

I don't think we need to manually copying these unchanged files into a custom directory. That's indeed the purpose of the `MANIFEST` file to specify which files outside the package src directory should be included in the distribution.

**Review by maliasadi — COMMENTED** [review]
Thanks @erick-xanadu for exploring different ways of getting rid of patches! Here with committing the auto-generated files into the repo and updating them manually when the ODS definition added/updated instead of using patches to apply changes instead, I want to make an argument for whether this approach is ever better than our current approach.

Imo, patches are inconvenience when building from src as applying them will add complexity to a os-agnostic build process and in the long run managing patches across different versions of LLVM/MLIR and other associated dependencies may be more challenging. On the other hand, copying 3rd-party (auto-)generated files will add more headache regarding the code maintenance than just using patches. Further, we have to deal manually with any upstream changes to the ODS definitions more often. What do you think?

Let's also tag @mlxd in this thread as he's also opposed to using patches.

File: Makefile [hunk 1]
```diff
@@ -98,12 +98,6 @@ wheel:
 	# Copy enzyme to frontend
 	cp --dereference $(ENZYME_BUILD_DIR)/Enzyme/LLVMEnzyme-17.so $(MK_DIR)/frontend/catalyst/lib
 	# Copy mlir bindings to frontend/mlir_quantum
```
Comment (grwlf) [inline Makefile:100]: This comment looks no longer relevant

File: mlir/Makefile [hunk 2]
```diff
@@ -28,6 +28,20 @@ help:
 .PHONY: all
 all: llvm mhlo dialects enzyme
 
+.PHONY: patch
```
Comment (grwlf) [inline mlir/Makefile:31]: Could you please add the help (or doc?) string explaining when to call this rule?

**Review by grwlf — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
Comment (erick-xanadu) [reply Makefile:100]:
```suggestion
```

File: mlir/Makefile (hunk 2 above)
Comment (erick-xanadu) [reply mlir/Makefile:31]: I'll wait until the discussion for this branch is finished to update the docs, but thanks for pointing it out!

**Comment by mlxd** [comment]
> Thanks @erick-xanadu for exploring different ways of getting rid of patches! Here with committing the auto-generated files into the repo and updating them manually when the ODS definition added/updated instead of using patches to apply changes instead, I want to make an argument for whether this approach is ever better than our current approach.
> 
> Imo, patches are inconvenience when building from src as applying them will add complexity to a os-agnostic build process and in the long run managing patches across different versions of LLVM/MLIR and other associated dependencies may be more challenging. On the other hand, copying 3rd-party (auto-)generated files will add more headache regarding the code maintenance than just using patches. Further, we have to deal manually with any upstream changes to the ODS definitions more often. What do you think?
> 
> Let's also tag @mlxd in this thread as he's also opposed to using patches.

From my side there is still a lot for me to catch up on in here. Though, I think the direct use of the autogen'd files to be the least bad option --- I think there is no perfect solution here right now, but as a long term strategy this process will likely be improved by directly engaging with JAX/TF/LLVM authors and helping shape the ecosystem to support our needs.

I'll try a pass on this in the coming week, but when @erick-xanadu is back we can try to identify a strategy that fixes things through direct engagement with the other projects. I think having some steering council discussions, or similar approach, may be the long term solution here.

Failing that, we can follow-suit from the other orgs and fork the required dependencies to add the changes we require. Again, not idea, but if it helps us to manage the libs end-to-end it may be the best option (e.g. https://github.com/ROCmSoftwarePlatform/tensorflow-upstream. If we need full-control of the associated libs, another option is to provide docker images with all changes as needed, and let that be the delivery mechanism. But let's discuss this another time.

**Comment by erick-xanadu** [comment]
@dime10 Regarding checking in the runtime module, I think it is better to include it. This will remove the "INSTALLED" global variable.

**Comment by erick-xanadu** [comment]
@dime10 since you are taking over and using a different strategy, I'm also closing this PR.

**Pull request closed without merging**